action/protocol/execution/evm/evm.go

```diff
@@ -287,12 +287,13 @@ func ExecuteContract(
 	}
 	stateDB.clear()
 
-	if featureCtx.SetRevertMessageToReceipt && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrExecutionReverted) && retval != nil && bytes.Equal(retval[:4], _revertSelector) {
-		// in case of the execution revert error, parse the retVal and add to receipt
-		data := retval[4:]
-		msgLength := byteutil.BytesToUint64BigEndian(data[56:64])
-		revertMsg := string(data[64 : 64+msgLength])
-		receipt.SetExecutionRevertMsg(revertMsg)
+	if featureCtx.SetRevertMessageToReceipt && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrExecutionReverted) && retval != nil {
+		reason, errUnpack := abi.UnpackRevert(retval)
+		//we need to catch custom errors
+		if errUnpack != nil {
+			reason = "execution reverted"
+		}
+		receipt.SetExecutionRevertMsg(reason)
```

Comment:
will cause HF

Reply:
https://github.com/ethereum/go-ethereum/blob/a36c68f12cdc45f52184b563dab475f340e00dd7/accounts/abi/bind/backends/simulated.go#L403
keep same format with geth?

Reply:
1. here is only to capture custom errors while keeping the logic unchanged as before. 

2. geth’s handling has been moved to the API layer, and making changes here would cause HF, necessitating significant code adjustments.
https://github.com/iotexproject/iotex-core/pull/3920/files#diff-6dc4e8a6418cb4f7e5b7da626b745fb04ff21d7c37b77d883b8c8b9ab7aacfcbR403

Reply:
geth code
```
typ, err := NewType("string", "", nil)
	if err != nil {
		return "", err
	}
	unpacked, err := (Arguments{{Type: typ}}).Unpack(data[4:])
	if err != nil {
		return "", err
	}
	return unpacked[0].(string), nil
```
I am not 100% sure, if that will be exactly same as our code
```
data := retval[4:]
msgLength := byteutil.BytesToUint64BigEndian(data[56:64])
revertMsg := string(data[64 : 64+msgLength])
```

Reply:
100% sure. if not, the check of ExecutionRevertMsg in tests will not passed

Reply:
```
if fCtx.CorrectCustomRevert { 
    your code in PR
} else if bytes.Equal(retval[:4], _revertSelector) {
    current code
}

```